### PR TITLE
Fix - should merge link sources from overwrite document

### PR DESCRIPTION
**Pull request by hellosnow**

### CODE REVIEW

File: src/Microsoft.DocAsCode.Build.Common/Reference/ApplyOverwriteDocument.cs [hunk 1]
```diff
@@ -92,6 +92,8 @@ select new
                 Merge(vm, ovm, ovms[0].fileModel);
                 pair.model.LinkToUids = pair.model.LinkToUids.Union(od[0].LinkToUids);
                 pair.model.LinkToFiles = pair.model.LinkToFiles.Union(od[0].LinkToFiles);
+                pair.model.FileLinkSources = pair.model.FileLinkSources.Union(od[0].FileLinkSources).ToImmutableDictionary();
```
Comment (vwxyzh) [inline src/Microsoft.DocAsCode.Build.Common/Reference/ApplyOverwriteDocument.cs]: apply for each overwrite document.

File: src/Microsoft.DocAsCode.Build.Common/Reference/OverwriteDocumentModel.cs [hunk 2]
```diff
@@ -57,6 +58,20 @@ public class OverwriteDocumentModel
         [JsonIgnore]
         public HashSet<string> LinkToUids { get; set; } = new HashSet<string>();
 
+        /// <summary>
+        /// Link sources information for file
+        /// </summary>
+        [YamlIgnore]
+        [JsonIgnore]
+        public ImmutableDictionary<string, ImmutableList<LinkSourceInfo>> FileLinkSources { get; set; } = ImmutableDictionary<string, ImmutableList<LinkSourceInfo>>.Empty;
```
Comment (vwxyzh) [inline src/Microsoft.DocAsCode.Build.Common/Reference/OverwriteDocumentModel.cs]: Immutable type cannot serialize in binary format, it is depended by 32-bit process.

File: src/Microsoft.DocAsCode.Build.Common/Reference/ApplyOverwriteDocument.cs (hunk 1 above)
Comment (hellosnow) [inline src/Microsoft.DocAsCode.Build.Common/Reference/ApplyOverwriteDocument.cs]: good catch~

---
In reply to: [99538362](https://github.com/dotnet/docfx/pull/1269#discussion_r99538362) [](ancestors = 99538362)

File: src/Microsoft.DocAsCode.Build.Common/Reference/ApplyOverwriteDocument.cs [hunk 3]
```diff
@@ -90,8 +90,13 @@ select new
             {
                 var vm = pair.item;
                 Merge(vm, ovm, ovms[0].fileModel);
-                pair.model.LinkToUids = pair.model.LinkToUids.Union(od[0].LinkToUids);
-                pair.model.LinkToFiles = pair.model.LinkToFiles.Union(od[0].LinkToFiles);
+                foreach (var overwrite in overwrites)
```
Comment (superyyrrzz) [inline src/Microsoft.DocAsCode.Build.Common/Reference/ApplyOverwriteDocument.cs]: >overwrites [](start = 42, length = 10)

Should this be ovms?

File: src/Microsoft.DocAsCode.Build.Common/Reference/ApplyOverwriteDocument.cs (hunk 1 above)
Comment (hellosnow) [inline src/Microsoft.DocAsCode.Build.Common/Reference/ApplyOverwriteDocument.cs]: I'll fix it in next separate review, to distinguash with this one.

---
In reply to: [99538801](https://github.com/dotnet/docfx/pull/1269#discussion_r99538801) [](ancestors = 99538801,99538362)

**Comment by superyyrrzz** [comment]
:shipit:

File: src/Microsoft.DocAsCode.Build.Common/Reference/BuildReferenceDocumentBase.cs [hunk 4]
```diff
@@ -39,6 +40,8 @@ protected virtual void BuildOverwrite(IHostService host, FileModel model)
             model.Content = overwrites;
             model.LinkToFiles = overwrites.SelectMany(o => o.LinkToFiles).ToImmutableHashSet();
             model.LinkToUids = overwrites.SelectMany(o => o.LinkToUids).ToImmutableHashSet();
+            model.FileLinkSources = overwrites.SelectMany(o => o.FileLinkSources).Select(i => new KeyValuePair<string, ImmutableList<LinkSourceInfo>>(i.Key, i.Value.ToImmutableList())).ToImmutableDictionary();
```
Comment (superyyrrzz) [inline src/Microsoft.DocAsCode.Build.Common/Reference/BuildReferenceDocumentBase.cs]: `ToImmutableDictionary()` fails when `Select()` generates duplicated keys.

File: src/Microsoft.DocAsCode.Build.Common/Reference/ApplyOverwriteDocument.cs [hunk 5]
```diff
@@ -90,8 +90,12 @@ select new
...
                 var vm = pair.item;
                 Merge(vm, ovm, ovms[0].fileModel);
-                pair.model.LinkToUids = pair.model.LinkToUids.Union(od[0].LinkToUids);
-                pair.model.LinkToFiles = pair.model.LinkToFiles.Union(od[0].LinkToFiles);
+
+                // TODO: union from overwrites with same UID
+                pair.model.LinkToUids = pair.model.LinkToUids.Union(overwrites[0].LinkToUids);
+                pair.model.LinkToFiles = pair.model.LinkToFiles.Union(overwrites[0].LinkToFiles);
+                pair.model.FileLinkSources = pair.model.FileLinkSources.Union(overwrites[0].FileLinkSources).ToImmutableDictionary();
+                pair.model.UidLinkSources = pair.model.UidLinkSources.Union(overwrites[0].UidLinkSources).ToImmutableDictionary();
```
Comment (superyyrrzz) [inline src/Microsoft.DocAsCode.Build.Common/Reference/ApplyOverwriteDocument.cs]: `ToImmutableDictionary()` fails when `Select()` generates duplicated keys.

**Pull request merged**